### PR TITLE
debug_assert presence of `shard_id` tracing field

### PR DESCRIPTION
**Pull request by problame**
also:
fixes https://github.com/neondatabase/neon/issues/6638

### CODE REVIEW

File: pageserver/src/span.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#[cfg(not(debug_assertions))]
+pub(crate) fn debug_assert_current_span_has_tenant_id() {}
+
+#[cfg(debug_assertions)]
+#[track_caller]
+pub(crate) fn debug_assert_current_span_has_tenant_id() {
+    if let Err(missing) = check_fields_present!([&*TENANT_ID_EXTRACTOR, &*SHARD_ID_EXTRACTOR]) {
+        panic!("missing extractors: {missing:?}")
+    }
+}
```
Comment (koivunej) [inline pageserver/src/span.rs:23]:
```suggestion
#[track_caller]
pub(crate) fn debug_assert_current_span_has_tenant_id() {
    #[cfg(debug_assertions)]
    if let Err(missing) = check_fields_present!([&*TENANT_ID_EXTRACTOR, &*SHARD_ID_EXTRACTOR]) {
        panic!("missing extractors: {missing:?}")
    }
}
```

Is there a benefit in doing the two versions one?

**Review by koivunej — APPROVED** [review]
Approving anyways, this makes sense, I guess we both suggested this earlier.

I would like to get my #6621 merged before though :)

File: pageserver/src/span.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+#[cfg(debug_assertions)]
+static TENANT_ID_EXTRACTOR: once_cell::sync::Lazy<MultiNameExtractor<1>> =
+    once_cell::sync::Lazy::new(|| MultiNameExtractor::new("TenantId", ["tenant_id"]));
+#[cfg(debug_assertions)]
+static SHARD_ID_EXTRACTOR: once_cell::sync::Lazy<MultiNameExtractor<1>> =
+    once_cell::sync::Lazy::new(|| MultiNameExtractor::new("ShardId", ["shard_id"]));
+#[cfg(debug_assertions)]
+static TIMELINE_ID_EXTRACTOR: once_cell::sync::Lazy<MultiNameExtractor<1>> =
+    once_cell::sync::Lazy::new(|| MultiNameExtractor::new("TimelineId", ["timeline_id"]));
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/span.rs:12]: Also, multiname is no longer needed, these could just be `"tenant_id"` as in field keys.

File: pageserver/src/span.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/span.rs:23]: Didn't think about it. But, would the `track_caller` make a difference?

Comment (koivunej) [reply pageserver/src/span.rs:23]: No, it's effect is compile time only.

File: pageserver/src/span.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/span.rs:12]: https://github.com/neondatabase/neon/issues/6638

File: pageserver/src/span.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/span.rs:23]: The benefit of this suggestion doesn't warrant the time I spend clicking retry on the CI, sadly -,-

Comment (problame) [reply pageserver/src/span.rs:23]: Will shove this into https://github.com/neondatabase/neon/issues/6638

**Pull request merged**